src/Common/Config/ConfigProcessor.cpp

```diff
@@ -755,6 +843,11 @@ void ConfigProcessor::savePreprocessedConfig(const LoadedConfig & loaded_config,
     {
         LOG_WARNING(log, "Couldn't save preprocessed config to {}: {}", preprocessed_path, e.displayText());
     }
+
+#if USE_SSL
+    if (fs::path(preprocessed_path).filename() == "config.xml")
```

Comment:
SIlly question, but will this work with yaml configuration files?

Reply:
The question is interesting. 

I am trying to figure it out. It seems that in current code it cannot work with yaml files. 
But we can add support of yaml if needed.

In the integration test we have:
tests/integration/test_config_decryption/configs/config.xml

```xml
<clickhouse>
    <encryption_codecs>
        <aes_128_gcm_siv>
            <key_hex>00112233445566778899aabbccddeeff</key_hex>
        </aes_128_gcm_siv>
        <aes_256_gcm_siv>
            <key_hex>00112233445566778899aabbccddeeff00112233445566778899aabbccddeeff</key_hex>
        </aes_256_gcm_siv>
    </encryption_codecs>
    <max_table_size_to_drop encryption_codec="AES_128_GCM_SIV">96260000000B0000000000E8FE3C087CED2205A5071078B29FD5C3B97F824911DED3217E980C</max_table_size_to_drop>
    <max_partition_size_to_drop encryption_codec="AES_256_GCM_SIV">97260000000B0000000000BFFF70C4DA718754C1DA0E2F25FF9246D4783F7FFEC4089EC1CC14</max_partition_size_to_drop>
</clickhouse>
```

I used [ymltoxml](https://github.com/sarnold/ymltoxml) to convert it into yaml by command:
```bash
ymltoxml -i config.xml -o config.yaml
```

I got:
```yaml
clickhouse:
  encryption_codecs:
    aes_128_gcm_siv:
      key_hex: 00112233445566778899aabbccddeeff
    aes_256_gcm_siv:
      key_hex: 00112233445566778899aabbccddeeff00112233445566778899aabbccddeeff
  max_table_size_to_drop:
    '@encryption_codec': AES_128_GCM_SIV
    '#text': 96260000000B0000000000E8FE3C087CED2205A5071078B29FD5C3B97F824911DED3217E980C
  max_partition_size_to_drop:
    '@encryption_codec': AES_256_GCM_SIV
    '#text': 97260000000B0000000000BFFF70C4DA718754C1DA0E2F25FF9246D4783F7FFEC4089EC1CC14
```

The problem is that attributes via `@` as supported by ClickHouse. See the example of usage [here](https://github.com/ClickHouse/ClickHouse/blob/master/tests/integration/test_config_xml_yaml_mix/configs/config.d/mark_cache_size.yaml).

But text nodes via `#text` are not supported yet. See implementation of YAML parser: 
https://github.com/ClickHouse/ClickHouse/blob/bfe349a7e459a68bb074f141d07049c7a9c252ca/src/Common/Config/YAMLParser.cpp#L103
So we could implement support of `#text` in YAMLParser.cpp.

Reply:
@rschu1ze 
OK. I added support of text nodes in YAMLParser.cpp. Thus YAML configs can be decrypted too. Also I added a test for YAML configs decryption into my integration test.

In short I see there is no a single/standard approach about attributes and text nodes in YAML.

Consider the fragment of XML config:
```xml
  <max_table_size_to_drop encryption_codec="AES_128_GCM_SIV">96260000000B0000000000E8FE3C087CED2205A5071078B29FD5C3B97F824911DED3217E980C</max_table_size_to_drop>
```

In this fragment:
- `max_table_size_to_drop` is an element node.
- `encryption_codec` is an attribute node
- `96260000000B0000000000E8FE3C087CED2205A5071078B29FD5C3B97F824911DED3217E980C` is a text node.

I tried different online tools they convert XML to YAML differently.

https://www.site24x7.com/tools/xml-to-yaml.html
```yaml
  max_table_size_to_drop:
    "-encryption_codec": AES_128_GCM_SIV
    "#text": 96260000000B0000000000E8FE3C087CED2205A5071078B29FD5C3B97F824911DED3217E980C
```

https://www.atatus.com/tools/xml-to-yaml
```yaml
    max_table_size_to_drop: 
      encryption_codec: "AES_128_GCM_SIV"
      $t: "96260000000B0000000000E8FE3C087CED2205A5071078B29FD5C3B97F824911DED3217E980C"
```

https://beautifytools.com/xml-to-yaml-converter.php 
```yaml
 max_table_size_to_drop: 
  _encryption_codec: AES_128_GCM_SIV
  __text: 96260000000B0000000000E8FE3C087CED2205A5071078B29FD5C3B97F824911DED3217E980C
  toString: 
```

https://www.anyjson.in/xml-to-yaml
```yaml
  max_partition_size_to_drop:
    '#text': 97260000000B0000000000BFFF70C4DA718754C1DA0E2F25FF9246D4783F7FFEC4089EC1CC14
    '@encryption_codec': AES_256_GCM_S
```

This last result is the same as produced by mentioned about [ymltoxml](https://github.com/sarnold/ymltoxml) module.
And as far as we already represent attributes as nodes starting with `@` I implemented support of text nodes as `#text`.

Reply:
About the XML/YAML relationship: The authoritative guide (at least as far as I see) is https://clickhouse.com/docs/en/operations/configuration-files . It says about XML attributes:

```
If you want to write an attribute for a Sequence or Map node, you should use a @ prefix before the attribute key. Note, that @ is reserved by YAML standard, so you should also to wrap it into double quotes:

map:
  "@attr1": value1
  "@attr2": value2
  key: 123
```

which is close to the last of your examples but not 100% the same. Could you make your YAML test check this syntax?

Reply:
For this code `key` is not the text node. But it's the element subnode. Because it's "translated" into this code according to documentation:
```xml
<map attr1="value1" attr2="value2">
    <key>123</key>
</map>
```
While we need it to be translated into:
```xml
<map attr1="value1" attr2="value2">
123
</map>
```

According to the documentation probably the sequence can do it the way we need it:

```yaml
seq:
  - "@attr1": value1
  - "@attr2": value2
  - 123
```

```xml
<seq attr1="value1" attr2="value2">123</seq>
```

BTW in my correction of YAMLParser.cpp there no exception throwing for the case when several `#text` are specified. For example this way:

```yaml
  max_table_size_to_drop:
    '#text': SOMETHING1
    '@encryption_codec': AES_128_GCM_SIV
    '#text': SOMETHING2
````
1. Should we ignore it? 
2. Or should we throw an exception when we see `#text` the second time in config? 
3. Or should we silently replace old value  SOMETHING1 by new value SOMETHING2 when parsing?

Reply:
Yes, according to the docs, the YAML would need to contain a sequence of the form

```yaml
seq:
  - "@attr1": value1
  - "@attr2": value2
  - 123
```

instead of a map.

> Should we ignore it?
> Or should we throw an exception when we see #text the second time in config?
> Or should we silently replace old value SOMETHING1 by new value SOMETHING2 when parsing?

Assuming the user writes ClickHouse-specific YAML instead of YAML generated by the XML-to-YAML translators, do we need any modifications in the YAML parser at all or does the existing code work out of the box? (if it doesn't then the docs currently lie).

Besides that, I propose to implement the same behavior which exists in XML if the same tag appears multiple times, e.g.

```xml
<some_param>42</some_param>
<some_param>53</some_param>
```

Reply:
@rschu1ze I think the documentation is incorrect about usage the sequence with YAML. 
I tried the following:

If config.yaml file has:

```yaml
encryption_codecs:
  aes_128_gcm_siv:
    key_hex: 00112233445566778899aabbccddeeff
  aes_256_gcm_siv:
    key_hex: 00112233445566778899aabbccddeeff00112233445566778899aabbccddeeff
max_table_size_to_drop:
  - '@encryption_codec': AES_128_GCM_SIV
  - 96260000000B0000000000E8FE3C087CED2205A5071078B29FD5C3B97F824911DED3217E980C
max_partition_size_to_drop:
  - 97260000000B0000000000BFFF70C4DA718754C1DA0E2F25FF9246D4783F7FFEC4089EC1CC14
  - '@encryption_codec': AES_256_GCM_SIV
```

Then it created config-preprocessed.xml with contents:

```xml
    <encryption_codecs>
        <aes_128_gcm_siv>
            <key_hex>00112233445566778899aabbccddeeff</key_hex>
        </aes_128_gcm_siv>
        <aes_256_gcm_siv>
            <key_hex>00112233445566778899aabbccddeeff00112233445566778899aabbccddeeff</key_hex>
        </aes_256_gcm_siv>
    </encryption_codecs>
    <max_table_size_to_drop encryption_codec="AES_128_GCM_SIV"/>
    <max_table_size_to_drop>96260000000B0000000000E8FE3C087CED2205A5071078B29FD5C3B97F824911DED3217E980C</max_table_size_to_drop>
    <max_partition_size_to_drop>97260000000B0000000000BFFF70C4DA718754C1DA0E2F25FF9246D4783F7FFEC4089EC1CC14</max_partition_size_to_drop>
    <max_partition_size_to_drop encryption_codec="AES_256_GCM_SIV"/>
```

You see it's not just about the encryption elements this is the way YAMLParser.cpp works in master. Thus we cannot use the sequences.

What do we do then?

Reply:
Right now I implemented it this way (way number two): YAMLParser will throw an exception if it find second text node for an element node.